pkg/apis/config/settings/settings.go

```diff
@@ -0,0 +1,126 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package settings
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/samber/lo"
+	v1 "k8s.io/api/core/v1"
+	"knative.dev/pkg/configmap"
+
+	"github.com/aws/karpenter-core/pkg/apis/config"
+)
+
+var ContextKey = Registration
+
+var Registration = &config.Registration{
+	ConfigMapName: "karpenter-global-settings",
+	Constructor:   NewSettingsFromConfigMap,
+	DefaultData:   lo.Must(defaultSettings.Data()),
+}
+
+var defaultSettings = Settings{}
+
+type Settings struct {
+	Tags map[string]string
+}
+
+func (s Settings) UnmarshalJSON(raw []byte) error {
+	d := map[string]string{}
+	if err := json.Unmarshal(raw, &d); err != nil {
+		return err
+	}
+	if err := AsMap("aws.tags", &s.Tags)(d); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (s Settings) MarshalJSON() ([]byte, error) {
+	d := map[string]string{}
+	if err := FromMap(s.Tags)("aws.tags", &d); err != nil {
+		return nil, err
+	}
+	return json.Marshal(d)
+}
+
+func (s Settings) Data() (map[string]string, error) {
+	d := map[string]string{}
+
+	if err := json.Unmarshal(lo.Must(json.Marshal(defaultSettings)), &d); err != nil {
+		return d, fmt.Errorf("unmarshalling json data, %w", err)
+	}
+	return d, nil
+}
+
+// NewSettingsFromConfigMap creates a Settings from the supplied ConfigMap
+func NewSettingsFromConfigMap(cm *v1.ConfigMap) (Settings, error) {
+	s := defaultSettings
+
+	if err := configmap.Parse(cm.Data,
+		AsMap("aws.tags", &s.Tags),
+	); err != nil {
+		// Failing to parse means that there is some error in the Settings, so we should crash
+		panic(fmt.Sprintf("parsing config data, %v", err))
+	}
+	return s, nil
+}
+
+func ToContext(ctx context.Context, s Settings) context.Context {
```

Comment:
WDYT about making a generic for this? We could use a `*new(T)` for the context key. You have to make sure the injection is a singleton, but should work for most cases.

Reply:
I messed with it a bit and I'm not convinced I like it. The code seems to become less idiomatic like


```go
context.FromContext[settings.Settings](ctx).BatchIdleDuration.Duration

settings.FromContext(ctx).BatchIdleDuration.Duration
```

I could maybe go either way, but I'm not seeing a huge benefit to making it generic and the type checking actually becomes more confusing in a lot of ways